File: leetcode/30.串联所有单词的子串.py
from typing import List

# 写的是对的，但是会超时...


#
# @lc app=leetcode.cn id=30 lang=python3
#
# [30] 串联所有单词的子串
#


# @lc code=start
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        word_len = len(words[0])
        words_list_len = len(words)
        s_len = len(s)
        words_set = {x: words.count(x) for x in words}
        start = 0
        end = 0
        cnt = 0
        res = []
        if s_len < word_len * words_list_len:
            return []
        exist_set = {}
        while end < s_len:
            temp = end
            end += word_len
            cur_str = s[temp:end]
            # print(f"cur_str: {cur_str}")
            # print(f"exist_set: {exist_set}, start: {start}")
            if cur_str not in words_set or (
                cur_str in exist_set and exist_set[cur_str] >= words_set[cur_str]
            ):
                start += 1
                end = start
                cnt = 0
                exist_set.clear()
            else:
                cnt += 1
                if cur_str in exist_set:
                    exist_set[cur_str] += 1
                else:
                    exist_set[cur_str] = 1
                if cnt >= words_list_len:
                    res.append(start)
                    start += 1
                    end = start
                    cnt = 0
                    exist_set.clear()
        return res

# @lc code=end

s = Solution()
print(s.findSubstring("barfoothefoobarman", ["foo", "bar"]))
print(s.findSubstring("wordgoodgoodgoodbestword", ["word", "good", "best", "word"]))
print(s.findSubstring("barfoofoobarthefoobarman", ["bar", "foo", "the"]))
print(s.findSubstring("wordgoodgoodgoodbestword", ["word", "good", "best", "good"]))

File: leetcode/30.串联所有单词的子串.rs
/*
 * @lc app=leetcode.cn id=30 lang=rust
 *
 * [30] 串联所有单词的子串
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn find_substring(s: String, words: Vec<String>) -> Vec<i32> {
        let mut result = Vec::new();
        if words.is_empty() {
            return result;
        }
        
        let word_len = words[0].len();
        let total_words = words.len();
        let total_len = word_len * total_words;
        let s_len = s.len();
        
        if s_len < total_len {
            return result;
        }
        
        let mut target_map: HashMap<&str, usize> = HashMap::new();
        for word in &words {
            *target_map.entry(word.as_str()).or_insert(0) += 1;
        }
        
        for i in 0..word_len {
            let mut current_map = HashMap::new();
            let mut left = i;
            let mut count = 0;
            
            for j in (i..=s_len - word_len).step_by(word_len) {
                let current_word = &s[j..j + word_len];
                
                if !target_map.contains_key(current_word) {
                    current_map.clear();
                    count = 0;
                    left = j + word_len;
                    continue;
                }
                
                *current_map.entry(current_word).or_insert(0) += 1;
                count += 1;
                
                while *current_map.get(current_word).unwrap() > *target_map.get(current_word).unwrap() {
                    let left_word = &s[left..left + word_len];
                    *current_map.get_mut(left_word).unwrap() -= 1;
                    count -= 1;
                    left += word_len;
                }
                
                if count == total_words {
                    result.push(left as i32);
                    let left_word = &s[left..left + word_len];
                    *current_map.get_mut(left_word).unwrap() -= 1;
                    count -= 1;
                    left += word_len;
                }
            }
        }
        
        result
    }
}
// @lc code=end

